### PR TITLE
Rewrite without dolittle

### PR DESCRIPTION
**Pull request by rafaelschlatter**
- replaced old with new framework
- removed `PingTester.cs` and `BufferReader.cs` (code was duplicated)
- removed `Build`, ´bin`  and `obj` folders
- added GitHub actions
- added .sln file
- updated readme
- added .gitignore & .dockerignore

I am uncertain about these things:
- the Dockerfile has an if statement based on `CONFIGURATION`, in the GitHub action, I use `Release` to build the image
- For the `Buffer` and `Ping` task in `Connector.cs` lines 42 and 43, is there a difference between using `await Task.Completed`, and `await bufferTask` and `await pingTask`? I want that we see excpetions in the the tasks on the main thread, which I think is a not the case with `Task.Run()`, but I am not sure.

### CODE REVIEW

**Comment by srudaa** [comment]
> * For the `Buffer` and `Ping` task in `Connector.cs` lines 42 and 43, is there a difference between using `await Task.Completed`, and `await bufferTask` and `await pingTask`? I want that we see excpetions in the the tasks on the main thread, which I think is a not the case with `Task.Run()`, but I am not sure.

@thotokraa could you answer this?

**Comment by thotokraa** [comment]
> 
> 
> > * For the `Buffer` and `Ping` task in `Connector.cs` lines 42 and 43, is there a difference between using `await Task.Completed`, and `await bufferTask` and `await pingTask`? I want that we see excpetions in the the tasks on the main thread, which I think is a not the case with `Task.Run()`, but I am not sure.
> 
> @thotokraa could you answer this?

`await Task.CompletedTask` is quite literally just a constant representing a task that has already completed. It is only used to trick the compiler to make the async function compile, as any async function _requires_ an await statement inside it, otherwise it won't compile. No waiting will be enforced by this function, so declaring two tasks as variables, then awaiting for CompletedTask won't do what you think it does. What you'll want to use is `await Task.WhenAll(bufferTask, pingTask);`

If any exceptions happens inside a running async function, the exception will be thrown in the main thread when calling `.Wait()` on the task. [Check out this link to see what I mean](https://stackoverflow.com/questions/5383310/catch-an-exception-thrown-by-an-async-void-method)

File: Source/Connector.cs [hunk 1]
```diff
@@ -40,19 +38,17 @@ public Connector(ILogger logger, ConnectorConfiguration configuration)
...
-            Task.Run(Buffer);
-            Task.Run(Ping);
+            Task bufferTask = Buffer();
+            Task pingTask = Ping();
+            await Task.CompletedTask;
         }
-        /// <inheritdoc/>
 
-        Task Buffer()
+        private Task Buffer()
```
Comment (thotokraa) [inline Source/Connector.cs:51]: The task should also be async.

File: Source/Connector.cs [hunk 2]
```diff
@@ -61,50 +57,64 @@ Task Buffer()
...
+
+                    SendDatapoint(bufferSize);
                 }
                 catch (Exception ex)
                 {
                     _logger.Error(ex, "Error while trying to get buffer");
                 }
-            }
 
+                Thread.Sleep(_configuration.Sampling);
```
Comment (thotokraa) [inline Source/Connector.cs:76]: This sleep function will sleep the entire thread, which includes all other tasks running on the thread. Please use `Task.Delay(...)` instead.

File: Source/Connector.cs [hunk 3]
```diff
@@ -61,50 +57,64 @@ Task Buffer()
...
+            }
         }
 
-        /// <inheritdoc/>
-
-        Task Ping()
+        private Task Ping()
```
Comment (thotokraa) [inline Source/Connector.cs:80]: This task should be async.

File: Source/Connector.cs [hunk 4]
```diff
@@ -61,50 +57,64 @@ Task Buffer()
...
+                    };
+
+                    SendDatapoint(pingReply);
                 }
                 catch (Exception ex)
                 {
                     _logger.Error(ex, "Error while trying to ping");
                 }
+
                 Thread.Sleep(_configuration.Sampling);
```
Comment (thotokraa) [inline Source/Connector.cs:116]: You should use `Task.Delay(...)` instead when using tasks.

**Review by thotokraa — CHANGES_REQUESTED** [review]

**Comment by thotokraa** [comment]
> 
> 
> > > * For the `Buffer` and `Ping` task in `Connector.cs` lines 42 and 43, is there a difference between using `await Task.Completed`, and `await bufferTask` and `await pingTask`? I want that we see excpetions in the the tasks on the main thread, which I think is a not the case with `Task.Run()`, but I am not sure.
> > 
> > 
> > @thotokraa could you answer this?
> 
> `await Task.CompletedTask` is quite literally just a constant representing a task that has already completed. It is only used to trick the compiler to make the async function compile, as any async function _requires_ an await statement inside it, otherwise it won't compile. No waiting will be enforced by this function, so declaring two tasks as variables, then awaiting for CompletedTask won't do what you think it does. What you'll want to use is `await Task.WhenAll(bufferTask, pingTask);`
> 
> If any exceptions happens inside a running async function, the exception will be thrown in the main thread when calling `.Wait()` on the task. [Check out this link to see what I mean](https://stackoverflow.com/questions/5383310/catch-an-exception-thrown-by-an-async-void-method)

Looking at the code again, I see that neither `Buffer()` nor `Ping()` will ever return, which means that even if there was a `Wait()` function somewhere in the code to wait for `Connector.Run()` to finish, it would never happen. As I also see now, the connector is declared as a task through `ApplicationBuilder.WithTask<T>()`, so there would be no way to actually wait for the task to throw an exception. There's no way that I can see for handling these exceptions on the main thread, but why would you want this?

**Comment by rafaelschlatter** [comment]
> > > > * For the `Buffer` and `Ping` task in `Connector.cs` lines 42 and 43, is there a difference between using `await Task.Completed`, and `await bufferTask` and `await pingTask`? I want that we see excpetions in the the tasks on the main thread, which I think is a not the case with `Task.Run()`, but I am not sure.
> > > 
> > > 
> > > @thotokraa could you answer this?
> > 
> > 
> > `await Task.CompletedTask` is quite literally just a constant representing a task that has already completed. It is only used to trick the compiler to make the async function compile, as any async function _requires_ an await statement inside it, otherwise it won't compile. No waiting will be enforced by this function, so declaring two tasks as variables, then awaiting for CompletedTask won't do what you think it does. What you'll want to use is `await Task.WhenAll(bufferTask, pingTask);`
> > If any exceptions happens inside a running async function, the exception will be thrown in the main thread when calling `.Wait()` on the task. [Check out this link to see what I mean](https://stackoverflow.com/questions/5383310/catch-an-exception-thrown-by-an-async-void-method)
> 
> Looking at the code again, I see that neither `Buffer()` nor `Ping()` will ever return, which means that even if there was a `Wait()` function somewhere in the code to wait for `Connector.Run()` to finish, it would never happen. As I also see now, the connector is declared as a task through `ApplicationBuilder.WithTask<T>()`, so there would be no way to actually wait for the task to throw an exception. There's no way that I can see for handling these exceptions on the main thread, but why would you want this?

If we are not notified of these

File: Source/Connector.cs (hunk 1 above)
Comment (rafaelschlatter) [reply Source/Connector.cs]: Done

File: Source/Connector.cs (hunk 2 above)
Comment (rafaelschlatter) [reply Source/Connector.cs]: Done

**Review by thotokraa — APPROVED** [review]

**Pull request merged**